dimod/constrained/cyexpression.pyx

```diff
@@ -362,16 +362,73 @@ cdef class cyConstraintView(_cyExpression):
         self.constraint().mark_discrete(marker)
 
     def penalty(self):
-        penalty = self.constraint().penalty()
+        """The penalty type for a soft constraint.
 
-        if penalty == cppPenalty.LINEAR:
+        Returns:
+            If the constraint is soft, returns one of ``'linear'`` or ``'quadratic'``.
+            If the constraint is hard, returns ``None``.
+
+        """
+        constraint = self.constraint()
+        penalty = constraint.penalty()
+
+        if not constraint.is_soft():
+            return None
+        elif penalty == cppPenalty.LINEAR:
             return "linear"
         elif penalty == cppPenalty.QUADRATIC:
             return "quadratic"
         elif penalty == cppPenalty.CONSTANT:
-            return "constrant"
+            # user should never see this, but might as well be future proof
+            return "constant"
         else:
             raise RuntimeError("unexpected penalty")
 
+    def set_weight(self, weight, penalty='linear'):
+        """Set the weight of the constraint.
+
+        Args:
+            weight: Weight for a soft constraint.
+                Must be a positive number. If ``None`` or
+                ``float('inf')``, the constraint is hard.
+                In feasible solutions, all the model's hard constraints
+                must be met, while soft constraints might be violated to achieve
+                overall good solutions.
+
+            penalty: Penalty type for a soft constraint (a constraint with its
+                ``weight`` parameter set). Supported values are ``'linear'`` and
+                ``'quadratic'``.
+                ``'quadratic'`` is supported for a constraint with binary
+                variables only.
+
+        """
+        cdef bias_type _weight = float('inf') if weight is None else weight
+
+        if _weight <= 0:
+            raise ValueError("weight must be a positive number or None")
+
+        cdef cppConstraint[bias_type, index_type]* constraint = self.constraint()
+
+        cdef cppPenalty _penalty
+        if penalty == 'linear':
+           _penalty = cppPenalty.LINEAR
+        elif penalty == 'quadratic':
+            for i in range(constraint.num_variables()):
+                if (self.parent.cppcqm.vartype(constraint.variables()[i]) != cppVartype.BINARY
+                        and self.parent.cppcqm.vartype(constraint.variables()[i]) != cppVartype.SPIN):
```

Comment:
Perhaps this would be faster _and_ more readable:
```suggestion
                vartype = self.parent.cppcqm.vartype(constraint.variables()[i])
                if vartype not in (cppVartype.BINARY, cppVartype.SPIN):
```

Reply:
I had done it this way under the assumption that Cython wouldn't support the syntax you propose, but it generates
```C++
      switch (__pyx_v_vartype) {
        case dimod::BINARY:
        case dimod::SPIN:
        __pyx_t_2 = 0;
        break;
        default:
        __pyx_t_2 = 1;
        break;
      }
      __pyx_t_9 = (__pyx_t_2 != 0);
      if (unlikely(__pyx_t_9)) {
/* … */
      }
```
which is pretty damn neat.